shotover-proxy/src/codec/cassandra.rs

```diff
@@ -31,20 +33,58 @@ impl Default for CassandraCodec {
 impl CassandraCodec {
     pub fn new() -> CassandraCodec {
         CassandraCodec {
-            compression: Compression::None,
+            compression: Arc::new(RwLock::new(Compression::None)),
             messages: vec![],
             current_use_keyspace: None,
         }
     }
 }
 
+impl Codec for CassandraCodec {
+    fn clone_without_state(&self) -> Self {
+        Self {
+            compression: Arc::new(RwLock::new(Compression::None)),
+            messages: self.messages.clone(),
+            current_use_keyspace: self.current_use_keyspace.clone(),
+        }
+    }
+}
+
 impl CassandraCodec {
-    fn encode_raw(&mut self, item: CassandraFrame, dst: &mut BytesMut) {
-        let buffer = item.encode(self.compression);
-        if buffer.is_empty() {
-            info!("trying to send 0 length frame");
+    fn check_compression(&mut self, bytes: &BytesMut) -> Result<bool> {
```

Comment:
Does it make sense to store the compression state in the codec so we can just assume subsequent messages are all encoded?

Reply:
From the spec: 

> a server may only compress frames above a certain size at its
  discretion

https://github.com/apache/cassandra/blob/5f8270eb8670b7d2f68c8b1a1d791a8b2cb849a0/doc/native_protocol_v4.spec#L825

The compression used is agreed on in the startup message but subsequent messages won't always be encoded so we need to check whether the message is compressed each time, while saving the agreed upon compression.